shelley/chain-and-ledger/executable-spec/test/Generator/Utxo.hs

```diff
@@ -0,0 +1,142 @@
+                     UTxOState)
+import           Slot (Slot (..))
+import           STS.Ledger (LedgerEnv (..))
+import           Tx (pattern Tx, pattern TxBody, pattern TxOut)
+import           Updates (emptyUpdate)
+import           UTxO (pattern UTxO, balance, makeWitnessesVKey)
+
+-- | Generate a new transaction in the context of the LEDGER STS environment and state.
+--
+-- Selects unspent outputs and spends the funds on a some valid addresses.
+-- Also generates valid certificates.
+genTx :: LedgerEnv ShortHash MockDSIGN
+      -> (UTxOState, DPState)
+      -> KeyPairs
+      -> Gen Tx
+genTx (LedgerEnv _ slot _ pparams) (UTxOState utxo _ _ _, dpState) keys = do
+  keys' <- Gen.shuffle keys
+
+  -- inputs
+  (witnessedInputs, spendingBalance) <- pickSpendingInputs keys' utxo
+  let (inputs, spendWitnesses) = unzip witnessedInputs
+
+  -- output addresses
+  recipientAddrs <- genRecipients keys'
+
+  -- certificates
+  (certs, certWitnesses, deposits_) <- genDCerts keys' pparams dpState
+
+  -- attempt to make provision for certificate deposits (otherwise discard this generator)
+  when (spendingBalance < deposits_) Gen.discard
+  let balance_ = spendingBalance - deposits_
+
+  -- calc. fees and output amounts
+  let (fee, outputs) = calcFeeAndOutputs balance_ recipientAddrs
+
+  -- witnessed transaction
+  txBody <- genTxBody (Set.fromList inputs) outputs certs fee slot
+  let !wits = makeWitnessesVKey txBody (spendWitnesses ++ certWitnesses)
+      multiSig = Map.empty -- TODO @uroboros Generate multi-sig transactions
+
+  return (Tx txBody wits multiSig)
+
+-- | Generate a transaction body with the given inputs/outputs and certificates
+genTxBody
+  :: Set TxIn
+  -> [TxOut]
+  -> Seq DCert
+  -> Coin
+  -> Slot
+  -> Gen TxBody
+genTxBody inputs outputs certs fee slot = do
+  ttl <- Gen.integral $ Range.linear 1 100
+  return $ TxBody
+             inputs
+             outputs
+             certs
+             Map.empty -- TODO @uroboros generate withdrawals
+             fee
+             (slot + Slot ttl)
+             emptyUpdate -- TODO @uroboros generate updates
+
+-- | Calculate the fee and distribute the remainder of the balance
+-- to the given addresses (as transaction outputs)
+calcFeeAndOutputs
+  :: Coin
+  -> [Addr]
+  -> (Coin, [TxOut])
+calcFeeAndOutputs balance_ addrs =
+  ( fee + splitCoinRem
+  , (`TxOut` amountPerOutput) <$> addrs)
+  where
+    -- TODO @uroboros fee=0 works for now since PParams minFeeA/B == 0.
+    --      We need to instead add a "draft" TxBody as argument, which will
+    --      give a measure of the minimum fee for this transaction.
+    --      This estimated fee can then be used to create the actual TxBody,
+    --      with the new fee baked in.
+    fee = 0
+    -- split the available balance into equal portions (one for each address),
+    -- if there is a remainder, then add it to the fee.
+    balanceAfterFee = balance_ - fee
```

Comment:
is `balanceAfterFee` always non-negative?

Reply:
for now yes, with fee=0, but we need to do a symmetric subtract when we generate proper fees